### PR TITLE
[github] Use attributes as another way to provide metadata

### LINKED ISSUE #665: [github] Allow overriding Github params via attributes
Leftover work from #657
Blocked on #664

Now that we can provide metadata to the backend via attributes, we should also support both when this makes sense. For example, for Github we should be able to override the params that are extracted from the URL through these attributes, whether they are provided in the URL or not. 
Any data not provided should default to the existing defaults.

E.g. these should be all valid:

```html
<E mv-storage="https://github.com" mv-storage-username="leaverou" mv-storage-repo="mavoweb" mv-storage-path="data.json">
<E mv-storage="https://github.com/karger/otherrepo" mv-storage-username="leaverou" mv-storage-repo="mavoweb" mv-storage-path="data.json">
<E mv-storage="https://github.com/leaverou"mv-storage-repo="mavoweb" mv-storage-path="data.json">
<E mv-storage="https://github.com" mv-storage-username="leaverou" mv-storage-path="data.json">
```

This is trickier than it seems due to how `Mavo.Backend.Github.parseURL()` is implemented, where some values depend on other values.

### PR DESCRIPTION
**Pull request by DmitrySharabin**
Let's try to fix #665 (or at least begin iterating over it that I find beneficial). 😀

### CODE REVIEW

File: src/backend.github.js [hunk 1]
```diff
@@ -383,7 +383,13 @@ var _ = Mavo.Backend.register($.Class({
 				ret.filename = defaults.filename;
 			}
 
-			ret.filepath = path.join("/") || defaults.filepath || "";
+			// If an author provided backend metadata, use them
+			// since they have higher priority
+			for (const prop in ret) {
```
Comment (LeaVerou) [inline src/backend.github.js:388]: One issue with this is that it only iterates over properties already defined on `ret`, but a common case would be to define things via attributes and NOT in the URL, so they may not exist as keys on `ret`.

**Review by LeaVerou — COMMENTED** [review]
Hi Dmitry, thanks for working on this!
This largely works (see my one comment though).
The reason I didn't want to go that route is that it would be a better architecture if `parseURL()` just parses the URL like the function name implies, returns its components, and we override them as needed in the constructor. However, a big problem with that approach is that certain components depend on others, e.g. apiCall, path.
I'm not sure if the implementation simplicity is worth the architectural tradeoff.

File: src/backend.github.js [hunk 2]
```diff
@@ -20,7 +20,7 @@ var _ = Mavo.Backend.register($.Class({
 			filename: `${this.mavo.id}${extension}`
 		};
 
-		this.info = _.parseURL(this.source, this.defaults);
+		this.info = _.parseURL(this.source, this.defaults, o);
```
Comment (LeaVerou) [inline src/backend.github.js:23]: Do note there's another call to `parseURL` in `get()`. Do we need to modify that too?

**Comment by DmitrySharabin** [comment]
Hi Lea!

Thank you for the comments. I agree with your thoughts and share them.

> I'm not sure if the implementation simplicity is worth the architectural tradeoff.

Of course, it doesn't. Well, I continue working then. :)

**Comment by LeaVerou** [comment]
Btw, I did mean I wasn't sure. If anything, I was leaning towards that the tradeoff might be worth it. But if you want to look into it a bit more first, that's fine!

**Comment by DmitrySharabin** [comment]
Yes, I'd like to look into it a bit more. Every single time I learn a lot from these iterations. And I appreciate that. Thank you.

**Comment by DmitrySharabin** [comment]
Hi @LeaVerou,

Will you please give this PR another look when you have time? I made some adjustments and would love to have feedback from you.

Thanks!

**Comment by LeaVerou** [comment]
Thank you for working on this Dmitry. I will review properly in a bit.
Thought: I wonder if it would simplify things to express computed properties like `apiCall` as a writable accessor.
What do I mean by that, since you can't both specify accessors and `writable: true`? Something like this:

```js
Object.defineProperty(ret, "apiCall", {
	get () { 
		return /* expression */
	}, 
	set (v) {
		delete this.apiCall;
		this.apiCall = v;
	},
	configurable: true, 
	enumerable: true
});
```

**Comment by DmitrySharabin** [comment]
I saw a similar code in the Mavo code base, but, honestly, I’ve never used it myself. It’s not in my blood. Yet. :) Thank you for your suggestion (question). I’ll examine this object method on MDN (or elsewhere) and give it a try. I will be back soon. ;)

**Comment by DmitrySharabin** [comment]
@LeaVerou Thank you so much for your suggestion and the brilliant following blog post that helped me A LOT. If I got everything correctly, the code we have now is more maintainable, DRY, and without any doubt, more professional. 🤞🏻 The last word is up to you, of course. ☺️

Your help is priceless. I appreciate that. Thank you! 🙇

**Review by LeaVerou — APPROVED** [review]
LGTM. Nice work, Dmitry.

I think we should move the writable getters to a helper in `util.js` (or even one in Bliss), but we can do that via a separate PR later.

Next step would be to add some tests for this, both for specifying params via attributes that *aren't* present in the URL, as well as *overriding* params that are present.

**Comment by DmitrySharabin** [comment]
Thank you, Lea!

> Next step would be to add some tests for this, both for specifying params via attributes that aren't present in the URL, as well as overriding params that are present.

Yes, it's on my radar.

> I think we should move the writable getters to a helper in util.js (or even one in Bliss), but we can do that via a separate PR later.

Vote for Bliss. :) Would love to contribute to that awesome project too. ☺️

**Comment by LeaVerou** [comment]
> Vote for Bliss. :) Would love to contribute to that awesome project too. ☺️

Then hold off for a bit longer, I'm rewriting bliss to use ESM so until that ships, no new features.

**Comment by DmitrySharabin** [comment]
Deal! 🤝

**Pull request merged**